src/platform/packages/shared/kbn-esql-utils/src/utils/query_parsing_helpers.ts

```diff
@@ -147,3 +149,35 @@ export const getQueryColumnsFromESQLQuery = (esql: string): string[] => {
 
   return columns.map((column) => column.name);
 };
+/**
+ * This function is used to map the variables to the columns in the datatable
+ * @param esql:string
+ * @param variables:ESQLControlVariable[]
+ * @param columns:DatatableColumn[]
+ * @returns DatatableColumn[]
+ */
+export const mapVariableToColumn = (
+  esql: string,
+  variables: ESQLControlVariable[],
+  columns: DatatableColumn[]
+): DatatableColumn[] => {
+  if (!variables.length) {
+    return columns;
+  }
+  const { root } = parse(esql);
+  const usedVariablesInQuery = Walker.params(root).map((param) => param);
+
+  const uniqueVariablesInQyery = new Set<string>();
+  usedVariablesInQuery.forEach((variable) => {
+    uniqueVariablesInQyery.add(variable.text.replace('?', ''));
+  });
+
+  columns.map((column) => {
+    if (variables.some((variable) => variable.value === column.id)) {
+      const potentialColumnVariables = variables.filter((variable) => variable.value === column.id);
+      const variable = potentialColumnVariables.find((v) => uniqueVariablesInQyery.has(v.key));
+      column.variable = variable?.key ?? '';
+    }
```

Comment:
nit: we're iterating here multiple times for no reason.
1 Loop should be enough:
```ts
const variable = variables.find( variable => variable.value === column.id && uniqueVariablesInQyery.has(variable.key) );
column.variable = variable?.key ?? '';
```

Reply:
This creates a different behavior in some cases so I will prefer to keep what I have so far.

Reply:
What different behaviour?

Reply:
For example when I have a control `?agent_name `, make the change you propose and run `FROM logs* | WHERE agent.name == ?agent_name ` the table that is created is wrong

It produces this
<img width="358" alt="image" src="https://github.com/user-attachments/assets/b5dd0e1c-8189-45be-8fd8-26af9320f076" />

instead of
<img width="333" alt="image" src="https://github.com/user-attachments/assets/052b2c2e-a991-4104-8c97-f06e87eb94f9" />